allennlp/training/callback_trainer.py

```diff
@@ -234,7 +300,10 @@ def from_params(cls,  # type: ignore
         parameters = [[n, p] for n, p in model.named_parameters() if p.requires_grad]
         optimizer = Optimizer.from_params(parameters, params.pop("optimizer"))
 
-        model_save_interval = params.pop_float("model_save_interval", None)
+        if "moving_average" in params:
```

Comment:
It's weird to me that `MovingAverage` can't be a callback.

What about if all the callbacks were keyed by name and set as attributes (or even the trainer just has a "_has_moving_average" flag), so you could do the same kind of check in the other `Callbacks` which need to know if there is also a `MovingAverage` callback? Then they could just call the relevant callback and not have to have this in the Trainer. It's not a big deal but it just seems a little out of place.

Reply:
thanks for the prod, I figured out a way to do this where the trainer doesn't have to know about this at all.

basically the CallbackHandler maintains an index of callbacks by _type_, and then the Validate and Checkpoint callbacks can just ask the handler "do you have any moving average callbacks"

take a look

Reply:
actually, I changed my mind about that too, now any callback that wants to know about `MovingAverage` objects just iterates over the callbacks looking for them once at `TRAINING_START`